rfc/006-semantic-versioning.md

```diff
@@ -0,0 +1,76 @@
+3. External clients, for example, wallets and block explorers are those that perform any of the following:
+     - consume the RPC endpoints of nodes like `/block`
+     - subscribe to events that are streamed via websockets
+     - make queries to the indexer
+  
+    This set are concerned with chain upgrades which will impact their ability to query state and block data as well as broadcast transactions.
+  
+4. There is also a fourth emerging user: IBC module and relayers (although in many ways relayers can also be considered as external clients). The developers of IBC and consumers of their software are concerned about changes that may affect a chains ability to send arbitrary messages to another chain. Specifically, this can happen with a breaking change to the light client verification algorithm.  
+
+In a more broader sense, these four users are inextricably linked and their concerns reflect that of the end user of the applications. The major conclusion that this RFC makes is that **the ability for chains to provide continual service is more important than the actual upgrade burden put on the developers of these chains**. What that means is that although it may be burdensome for application developers to upgrade Tendermint versions, this is far outweighed by the pain node operators experience in upgrading, by the pain that external clients have in accommodating changes and ultimately the pain of end users who are benefiting from the application. Developers have all the time in world to make these changes whilst for other user groups, upgrading can be considered part of this real-time critical path.  
+
+## Proposal
+
+Tendermint will follow a same formatting structure of [SemVer](https://semver.org/) with a major, minor and patch version but with slight adjustments to what they encompass:
+
+A major version in Tendermint guarantees that for the entire cycle of that major version:
+
+- All blocks and state data in a blockchain can be queried. All headers can be verified even across minor version changes. Nodes can both block sync and state sync from genesis to the head of the chain.
+- Nodes in a network are able to communicate and perform BFT state machine replication so long as the agreed network version is the lowest of all nodes in a network. For example, nodes using version 1.5.x and 1.2.x can operate together so long as the network version is 1.2 or lower (but still within the 1.x range). This rule essentially captures the concept of network backwards compatibility.
+- An external client should not need to modify their code to interact with a node's RPC endpoints. New endpoints can be added but existing ones can not be altered or removed.
+- Migrations should be automatic. Upgrading of one node can happen asynchronously with respect to other nodes (although agreement of a network-wide upgrade will happen synchronously via consensus).
```

Comment:
I think we want migrations for both coordinated and rolling upgrades to be automatic. I think the only clause for this is if we find the migration to be something optional (an example of which is migrating the peer store which in the end we decided not to do).

Reply:
I see. The first sentence equates "automatic" with "asynchronous", while the second describes "network-wide" upgrades as having to be synchronous, so I thought they were meant to be an exception.

Are there two distinctions we need to make here, e.g., "synchronous vs. asynchronous" and "automatic vs. hands-on" or something like that?

Reply:
Right. I was associating automatic with not requiring manual intervention from the node operators although I'm also not sure if this is a gaurantee that we can really make.

Reply:
This seems like a choice we can make. Some options that occur to me (there may be more):

- Option 1: "We promise that any changes within a major version do not require manual upgrade of nodes in standard configuration."[1] Meaning: We will defer changes that _do_ require manual intervention to the next major version.
- Option 2: "We promise that there will be at most 1 change within a major version that requires manual upgrade of nodes in standard configuration."[1] Meaning: We will allow a budget of non-breaking manual upgrades within a major version.
- Option 3: "We will try really really hard to ensure that all changes within a major version can be upgraded without manual intervention for the majority of nodes." Meaning: Best-effort and we will try to provide tools, but we won't stop the world if we can't get everybody.

Given that we do not have a reliable customer list, (3) may be the only thing we can practically commit to. But from the user's perspective it isn't really a commitment they can rely on, since it allows us to throw up our hands at any time. To the extent that version promises are a means of building trust, we should choose cautiously.

[1] For some definition of "standard". Basically leaving an escape-hatch for weird and/or highly-customized nodes.